src/mediapipe_internal/mediapipegraphdefinition.cpp

```diff
@@ -380,6 +391,17 @@ std::string MediapipeGraphDefinition::getStreamName(const std::string& streamFul
     return EMPTY_STREAM_NAME;
 }
 
+Status MediapipeGraphDefinition::getPythonNodeState(const std::string& node_name, std::unique_ptr<NodeState>& nodeState) {
+    auto it = this->pythonNodeStates.find(node_name);
+    if (it == this->pythonNodeStates.end()) {
+        SPDLOG_LOGGER_DEBUG(modelmanager_logger, "Python node name: {} does not exist in graph: {}. ", node_name, this->name);
+        return StatusCode::PYTHON_NODE_NAME_DOES_NOT_EXISTS;
+    }
+
+    nodeState = std::move(it->second);
```

Comment:
This will segfault when used twice.

My suggestion is to define contract how long should node states live, why?
- will node states be removed when graph becomes retired? (what would happen to already running graphs?)
- is definition removed when graph is removed from config.json? (probably not?)
- will we remove node states only upon application close?

Depending on the answers given above, you should use unique_ptr/shared_ptr.

If node states will keep on living until the end of an application, it would be safe to keep node states in `unordered_map<key, unique_ptr<NodeState>>` and acquire pointer when neccessary:

```
Status MediapipeGraphDefinition::getPythonNodeState(const std::string& node_name, NodeState** nodeState) {
...
 *nodeState = &it->second;
 return status;
}
```

If we want to clear the state when graph becomes retired, we need to make sure graphs that are still running still have valid reference to the state. I would suggest unordered_map<key, shared_ptr<NodeState>> and:
```
Status MediapipeGraphDefinition::getPythonNodeState(const std::string& node_name, std::shared_ptr<NodeState>& nodeState) {
...
 nodeState = it->second;
 return status;
}
```

Reply:
I think that node state should get removed when the graph retires. If we don't remove graph definition then I think we should clear the map so we free all the resources.